libyul/optimiser/RedundantAssignEliminator.cpp

```diff
@@ -215,101 +81,79 @@ void RedundantAssignEliminator::operator()(Leave const&)
-	// TODO Perhaps it is better to just create a sorted list
-	// and then use insert(begin, end)
+	RedundantStoreBase::visit(_statement);
 
-	auto ita = _a.begin();
-	auto aend = _a.end();
-	auto itb = _b.begin();
-	auto bend = _b.end();
-
-	for (; itb != bend; ++ita)
-	{
-		if (ita == aend)
-			ita = _a.insert(ita, std::move(*itb++));
-		else if (ita->first < itb->first)
-			continue;
-		else if (itb->first < ita->first)
-			ita = _a.insert(ita, std::move(*itb++));
-		else
-		{
-			_conflictSolver(ita->second, std::move(itb->second));
-			++itb;
-		}
-	}
+	if (auto const* assignment = get_if<Assignment>(&_statement))
+		if (assignment->variableNames.size() == 1)
+			// Default-construct it in "Undecided" state if it does not yet exist.
+			m_stores[assignment->variableNames.front().name][&_statement];
 }
 
-void RedundantAssignEliminator::merge(TrackedAssignments& _target, TrackedAssignments&& _other)
+void RedundantAssignEliminator::shortcutNestedLoop(TrackedStores const& _zeroRuns)
 {
-	joinMap(_target, move(_other), [](
-		map<Assignment const*, State>& _assignmentHere,
-		map<Assignment const*, State>&& _assignmentThere
-	)
-	{
-		return joinMap(_assignmentHere, move(_assignmentThere), State::join);
-	});
+	// Shortcut to avoid horrible runtime:
+	// Change all assignments that were newly introduced in the for loop to "used".
+	// We do not have to do that with the "break" or "continue" paths, because
+	// they will be joined later anyway.
+	// TODO parallel traversal might be more efficient here.
+	for (auto& [variable, stores]: m_stores)
+		for (auto& assignment: stores)
+		{
+			auto zeroIt = _zeroRuns.find(variable);
+			if (zeroIt != _zeroRuns.end() && zeroIt->second.count(assignment.first))
+				continue;
+			assignment.second = State::Value::Used;
+		}
 }
 
-void RedundantAssignEliminator::merge(TrackedAssignments& _target, vector<TrackedAssignments>&& _source)
+void RedundantAssignEliminator::finalizeFunctionDefinition(FunctionDefinition const& _functionDefinition)
 {
-	for (TrackedAssignments& ts: _source)
-		merge(_target, move(ts));
-	_source.clear();
+	for (auto const& param: _functionDefinition.parameters)
+		finalize(param.name, State::Unused);
+	for (auto const& retParam: _functionDefinition.returnVariables)
+		finalize(retParam.name, State::Used);
 }
 
 void RedundantAssignEliminator::changeUndecidedTo(YulString _variable, RedundantAssignEliminator::State _newState)
 {
-	for (auto& assignment: m_assignments[_variable])
+	for (auto& assignment: m_stores[_variable])
 		if (assignment.second == State::Undecided)
 			assignment.second = _newState;
 }
 
 void RedundantAssignEliminator::finalize(YulString _variable, RedundantAssignEliminator::State _finalState)
 {
-	std::map<Assignment const*, State> assignments;
-	joinMap(assignments, std::move(m_assignments[_variable]), State::join);
-	m_assignments.erase(_variable);
+	std::map<Statement const*, State> stores;
+	util::joinMap(stores, std::move(m_stores[_variable]), State::join);
```

Comment:
Isn't this just
```
std::map<Statement const*, State> stores = std::move(m_stores[_variable]);
```
?

Reply:
Yes, not sure why I did it that way, maybe because of conformity? I can change it to use move it is easier to understand.